fe/fe-core/src/main/java/com/starrocks/scheduler/TaskRunManager.java

```diff
@@ -78,16 +78,23 @@ public SubmitResult submitTaskRun(TaskRun taskRun, ExecuteOption option) {
         return new SubmitResult(queryId, SubmitResult.SubmitStatus.SUBMITTED, taskRun.getFuture());
     }
 
-    public boolean killTaskRun(Long taskId) {
+    public boolean killTaskRun(Long taskId, boolean force) {
         TaskRun taskRun = taskRunScheduler.getRunningTaskRun(taskId);
         if (taskRun == null) {
             return false;
         }
-        taskRun.kill();
-        ConnectContext runCtx = taskRun.getRunCtx();
-        if (runCtx != null) {
-            runCtx.kill(false, "kill TaskRun");
-            return true;
+        try {
+            taskRun.kill();
+            ConnectContext runCtx = taskRun.getRunCtx();
+            if (runCtx != null) {
+                runCtx.kill(false, "kill TaskRun");
+                return true;
+            }
+        } finally {
+            // if it's force, remove it from running TaskRun map no matter it's killed or not
+            if (force) {
+                taskRunScheduler.removeRunningTask(taskRun.getTaskId());
```

Comment:
why do this can help to  avoid occupying resources too long?

Reply:
1. The running queue is limited by `Conf.task_runs_concurrency` which is 4 by default. If one task run is always in the running queue, the task run scheduler can not schedule other tasks for the limit.
2. If the task run is in the running queue,  other dependent  tasks (triggerred by this task) will also not be scheduled.